src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -38,60 +38,53 @@ internal static class RemoteHostAssetSerialization
 
             // special case
             if (checksums.Length == 0)
-            {
-                writer.WriteInt32(0);
                 return;
-            }
 
             if (singleAsset != null)
             {
-                writer.WriteInt32(1);
-                WriteAsset(writer, serializer, context, checksums[0], singleAsset, cancellationToken);
+                WriteAsset(writer, serializer, context, singleAsset, cancellationToken);
                 return;
             }
 
             Debug.Assert(assetMap != null);
-            writer.WriteInt32(assetMap.Count);
 
-            foreach (var (checksum, asset) in assetMap)
+            foreach (var checksum in checksums)
             {
-                WriteAsset(writer, serializer, context, checksum, asset, cancellationToken);
+                var asset = assetMap[checksum];
 
                 // We flush after each item as that forms a reasonably sized chunk of data to want to then send over the
                 // pipe for the reader on the other side to read.  This allows the item-writing to remain entirely
                 // synchronous without any blocking on async flushing, while also ensuring that we're not buffering the
                 // entire stream of data into the pipe before it gets sent to the other side.
+                WriteAsset(writer, serializer, context, asset, cancellationToken);
                 await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
             }
 
             return;
 
-            static void WriteAsset(ObjectWriter writer, ISerializerService serializer, SolutionReplicationContext context, Checksum checksum, SolutionAsset asset, CancellationToken cancellationToken)
+            static void WriteAsset(ObjectWriter writer, ISerializerService serializer, SolutionReplicationContext context, SolutionAsset asset, CancellationToken cancellationToken)
             {
                 Debug.Assert(asset.Kind != WellKnownSynchronizationKind.Null, "We should not be sending null assets");
-                checksum.WriteTo(writer);
                 writer.WriteInt32((int)asset.Kind);
 
                 // null is already indicated by checksum and kind above:
                 if (asset.Value is not null)
-                {
                     serializer.Serialize(asset.Value, writer, context, cancellationToken);
-                }
             }
         }
 
-        public static async ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(
-            PipeReader pipeReader, Checksum solutionChecksum, ISet<Checksum> checksums, ISerializerService serializerService, CancellationToken cancellationToken)
+        public static async ValueTask<ImmutableArray<object>> ReadDataAsync(
+            PipeReader pipeReader, Checksum solutionChecksum, ImmutableArray<Checksum> checksums, ISerializerService serializerService, CancellationToken cancellationToken)
```

Comment:
>ImmutableArray<Checksum> checksums [](http://example.com/codeflow?start=62&length=34)

This just needs a length, not the actual array anymore, right?

Reply:
nice catch!